src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,250 @@
+// featbin/extract-segments.cc
+
+// Copyright 2015  Tom Ko
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "feat/wave-reader.h"
+
+namespace kaldi {
+
+void FFTConvolution(Vector<BaseFloat> &x, Vector<BaseFloat> &h) {
+  int xdim = x.Dim();
+  int M = h.Dim();
+
+  KALDI_LOG << "M is " << M;
+
+  int N = 2;
+  while (N < 4 * M) {
+    N = N * 2;
+  }
+  KALDI_LOG << "Best N is " << N;
+
+  int L = N - M + 1;
+  KALDI_LOG << "L is " << L;
+  SplitRadixRealFft<BaseFloat> *srfft_;
+  srfft_ = new SplitRadixRealFft<BaseFloat>(N);
+
+  h.Resize(N, kCopyData);
+  srfft_->Compute(h.Data(), true);
+
+  Vector<float> temp_pad(M-1);
+  temp_pad.SetZero();
+  Vector<float> x_block_padded(N);
+
+  for (int po = 0; po < xdim; po += L) {
+    // get block of x
+    int remain = xdim - po;
+    int l = L;
+    if ( remain < l) l = remain;
+    SubVector<float> x_block(x, po, l);
+    x_block_padded.SetZero();
+    SubVector<float> ttt(x_block_padded, 0, l);
+    ttt.CopyFromVec(x_block);
+
+    srfft_->Compute(x_block_padded.Data(), true);
+
+    for (int i = 0; i < N / 2; i++) {
+      float tmp = x_block_padded(2*i);
+      x_block_padded(2*i) = tmp * h(2*i) - x_block_padded(2*i+1) * h(2*i+1);
+      x_block_padded(2*i+1) = tmp * h(2*i+1) + x_block_padded(2*i+1) * h(2*i);
+    }
+
+    srfft_->Compute(x_block_padded.Data(), false);
+    x_block_padded.Scale(1.0/N);
+
+    // combine the block
+    if (po + L < xdim) {       // current block is not the last block
+      for (int j = 0; j < L; j++)
+        x(po + j) = x_block_padded(j);
+      for (int j = 0; j < M-1; j++)
+        x(po + j) += temp_pad(j);
+      for (int j = 0; j < M-1; j++)
+        temp_pad(j) = x_block_padded(j + L);
+    } else {
+      for (int j = 0; j < xdim - po; j++)
+        x(po + j) = x_block_padded(j);
+      for (int j = 0; j < M-1 && j < xdim - po; j++)
+        x(po + j) += temp_pad(j);
+    }
+  }
+}
+
+void RepeatedlyAdd(Vector<BaseFloat> &x, Vector<BaseFloat> &h) {
```

Comment:
modify the signature to something like

void AddVectorsOfUnequalLength(const Vector<BaseFloat> &vector1, Vector<BaseFloat> *vector2 )